### ISSUE TITLE
library filenames

### ISSUE THREAD

**Issue by nschloe** [flame/blis#196]
Right now, BLIS inserts versions into file names à la
```
libblis-<something>.a
libblis-<something>.so
```
However, the generic versioned filenames for Unix libraries _append_ the version to shared library files like
```
libblas.a
libblas.so -> libblas.so.3.7.1
libblas.so.3 -> libblas.so.3.7.1
libblas.so.3.7.1
```
The intermediate `libblas.so.3` is for ABI-stable changes. Check your own `/usr/lib` for more examples.

Moreover, as you can see, static libraries are _never_ versioned. And why should they? The entire library is built into the executable anyways.

This should probably be adopted by BLIS as well.

**Comment by devinamatthews** [comment]
Just a quick nit: the x.y.z in libfoo.so.x.y.z is *not* quite the same as the library version. I understand the exact meaning of these numbers to be OS-dependent (see e.g. https://autotools.io/libtool/version.html). *BUT*, I think that since BLIS does not (yet) have multiple ABI versions that libblis.so -> libblis.so.0 -> libblis.so.0.0.0 would work.

libfoo-x.y.z.so is not unprecedented either, though.

**Comment by fgvanzee** [comment]
> Moreover, as you can see, static libraries are never versioned. And why should they? The entire library is built into the executable anyways.

I know you meant the question as more rhetorical, but there is actually an answer, and it is related to the inclusion of the architecture in the filename of the libraries. Sometimes, developers will be working on multiple hardware systems that all share the same home directory. If the developer is installing the libraries to his home directory, either out of preference (maybe he doesn't want to muck up his system directories), or because he lacks root access, including the architecture and version number can help with the management of the different library installs alongside one another.

In the case of an end-user who only ever wants one copy installed, I agree that the architecture and version number would be unnecessary for the static libraries (and maybe the shared as well, since as Devin pointed out, our version numbering is different from the .so version numbering).

**Comment by nschloe** [comment]
> including the architecture and version number can help with the management of the different library installs alongside one another.

This really sounds like a development thing to be then. Perhaps there could be `--devel` config flag that turns on this kind of stuff?

**Comment by fgvanzee** [comment]
@nschloe Yes, I think I can come up with a `configure` option that enables inclusion of architecture in the library names, and perhaps a separate option that enables inclusion of the version number. Thanks for the suggestion.

**Comment by fgvanzee** [comment]
@nschloe Okay, here's where I am. We now have a switch that toggles the build system between a "user" mode:
```
-rw-r--r-- 1 field dept 2979052 May  9 18:29 libblis.a
-rw-r--r-- 1 field dept 2185976 May  9 18:29 libblis.so
lrwxrwxrwx 1 field dept      10 May  9 18:29 libblis.so.0 -> libblis.so
lrwxrwxrwx 1 field dept      12 May  9 18:29 libblis.so.0.0.0 -> libblis.so.0
```
And a "developer" mode:
```
-rw-r--r-- 1 field dept 2979052 May  9 18:30 libblis-0.3.2-15-haswell.a
-rw-r--r-- 1 field dept 2185976 May  9 18:30 libblis-0.3.2-15-haswell.so
lrwxrwxrwx 1 field dept      26 May  9 18:30 libblis.a -> libblis-0.3.2-15-haswell.a
lrwxrwxrwx 1 field dept      27 May  9 18:30 libblis.so -> libblis-0.3.2-15-haswell.so
lrwxrwxrwx 1 field dept      10 May  9 18:30 libblis.so.0 -> libblis.so
lrwxrwxrwx 1 field dept      12 May  9 18:30 libblis.so.0.0.0 -> libblis.so.0
```
with the above being the output of `ls -l` on `LIBDIR` after installation, if it wasn't intuitively clear. Under this proposal, the build system would default to "user" mode.

Does this appear acceptable?

**Comment by devinamatthews** [comment]
Canonically `libblis.so` and `libblis.so.0` would be symlinks to `libblis.so.0.0.0`. Of course macOS is different, with either `libblis.dylib -> `libblis.0.dylib` and `libblis.0.dylib -> libblis.0.0.0.dylib` or with the former being `libblis.dylib -> libblis.0.0.0.dylib` as well. I don't think the numbers mean the same thing on macOS either but we don't actual have a version ABI so it doesn't matter.

**Comment by nschloe** [comment]
```
-rw-r--r-- 1 field dept 2979052 May  9 18:29 libblis.a
-rw-r--r-- 1 field dept 2185976 May  9 18:29 libblis.so
lrwxrwxrwx 1 field dept      10 May  9 18:29 libblis.so.0 -> libblis.so
lrwxrwxrwx 1 field dept      12 May  9 18:29 libblis.so.0.0.0 -> libblis.so.0
```
I would have expected `libblis.so.0.3.2` but looks good to me otherwise.

**Comment by fgvanzee** [comment]
> Canonically `libblis.so` and `libblis.so.0` would be symlinks to `libblis.so.0.0.0`

@devinamatthews Sorry about the symlinking fail. I hadn't taken a close enough look at what was conventional (or even what @nschloe used in his example) before deciding what to do. I'll fix it.

> I would have expected libblis.so.0.3.2 but looks good to me otherwise.

@nschloe Did you expect `libblis.so.0.3.2` because (a) it exactly matches the BLIS version number, or (b) you expected us to append the `3.2` of the BLIS version to the current `.so` major version number `.0` that is part of the `soname` (eg: `libblis.so.0`)?

**Comment by fgvanzee** [comment]
Updated output (developer mode):
```
lrwxrwxrwx 1 field dept       9 May 10 13:49 libblis-0.3.2-15-haswell.a -> libblis.a
lrwxrwxrwx 1 field dept      10 May 10 13:49 libblis-0.3.2-15-haswell.so -> libblis.so
-rw-r--r-- 1 field dept 2979052 May 10 13:49 libblis.a
lrwxrwxrwx 1 field dept      16 May 10 13:49 libblis.so -> libblis.so.0.0.0
lrwxrwxrwx 1 field dept      16 May 10 13:49 libblis.so.0 -> libblis.so.0.0.0
-rw-r--r-- 1 field dept 2185976 May 10 13:49 libblis.so.0.0.0
```
User mode would omit the version/configuration symlinks.

**Comment by fgvanzee** [comment]
I'm now leaning towards the following naming for user mode:
```
libblis.a
libblis.so -> libblis.so.0.0.0
libblis.so.0 -> libblis.so.0.0.0
libblis.so.0.0.0
```
and developer mode:
```
libblis-0.3.2-15-haswell.a
libblis-0.3.2-15-haswell.so
libblis.a -> libblis-0.3.2-15-haswell.a
libblis.so -> libblis.so.0.0.0
libblis.so.0 -> libblis.so.0.0.0
libblis.so.0.0.0 -> libblis-0.3.2-15-haswell.so

```

**Comment by devinamatthews** [comment]
I still don't quite see the utility of "developer mode" (over different out-of-tree builds/install prefixes), but I imagine you can do what you want with it since you/our group will probably be the only ones to use it.

**Comment by fgvanzee** [comment]
@devinamatthews I suppose you're right: the utility is limited to those who need to be able to *install* different configurations of BLIS to the same libdir, and now that I think about it, as I developer I rarely run `install` targets because the test infrastructure in BLIS is currently set up to link against the local (pre-installed) copy.

So I'll nix the developer mode. Less is more.

**Comment by fgvanzee** [comment]
@nschloe Look at commit b699bb1 and let me know if this will work for your needs.

**Comment by fgvanzee** [comment]
@nschloe Fabulous. Thanks so much for your contributions, Nico.

Would you mind if I linked to your packages from the main BLIS github page (`README.md`)?

**Comment by nschloe** [comment]
Not at all. I'll update the package once in a while.

**Comment by fgvanzee** [comment]
Great. We hope you stay involved in the future. :) Thanks again for helping make BLIS better and more usable.
